Use identifiers instead of raw pointers.

nihohit opened the pull request:
Instead of returning raw pointers, the atomic box now can return opaque identifiers of its internal state. This will ensure that the users will understand that the comparison results shouldn't be used as pointers, but are merely indicators of the the state of the atomic box.

ashdnazg commented:
I like this, but with two reservations:
1) The name identifier is very confusing. I thought about this more as an opaque ref, but refs might be confusing due to the meaning in rust.
2) You need to be able to get refs from other sources - for instance a box that you once had as is and not in an atomic box, or just comparing to a null pointer (a None optional box).

btw, since the pointer inside the identifier struct is not part of external interface, I think you can change it to be `mut` and save a bunch of casting.

nihohit commented:
1. AtomicBoxHandle?
2. Why? These values aren't in an atomic box, so the atomic box API doesn't need to take them into account. The core question for `compare_exchange` is "Has the value in the box changed?", since `compare_exchange` is an API that says "If the value in the box didn't change, set the new value in the box. Otherwise give me something to identify the current value". Non atomic box sources are irrelevant.
One thing we may consider is that operations like `swap` might return a pair of box and handle/identifier. We can't use the pointer as an identifier, since AFAIK nothing in the `Box` API promises us that the internal pointer is stable when moving.

nihohit commented:
Round:
1. Renamed AtomicBoxIdentifier to Handle
2. store / swap_mut methods now return a handle, so that users that set a value to an atomic box will automatically receive a handle to the current state of the box.
3. reverted load_raw in AtomicBox returning a handle. IMO we shouldn't expose the load_raw method on either public atomic.
4. reduced visibility of the base objects.
